sway-core/src/ir_generation/const_eval.rs

```diff
@@ -851,77 +890,87 @@ fn const_eval_intrinsic(
                 value: ConstantValue::Bool(val1 > val2),
-            }))
-        }
-        sway_ast::Intrinsic::Lt => {
-            let (ConstantValue::Uint(val1), ConstantValue::Uint(val2)) =
-                (&args[0].value, &args[1].value)
-            else {
-                unreachable!("Type checker allowed non integer value for LessThan")
-            };
-            Ok(Some(Constant {
+            })),
+            (ConstantValue::U256(val1), ConstantValue::U256(val2)) => Ok(Some(Constant {
+                ty: Type::get_bool(lookup.context),
+                value: ConstantValue::Bool(val1 > val2),
+            })),
+            _ => {
+                unreachable!("Type checker allowed non integer value for GreaterThan")
+            }
+        },
+        Intrinsic::Lt => match (&args[0].value, &args[1].value) {
+            (ConstantValue::Uint(val1), ConstantValue::Uint(val2)) => Ok(Some(Constant {
                 ty: Type::get_bool(lookup.context),
                 value: ConstantValue::Bool(val1 < val2),
-            }))
-        }
-        sway_ast::Intrinsic::AddrOf
-        | sway_ast::Intrinsic::PtrAdd
-        | sway_ast::Intrinsic::PtrSub
-        | sway_ast::Intrinsic::IsReferenceType
-        | sway_ast::Intrinsic::IsStrType
-        | sway_ast::Intrinsic::Gtf
-        | sway_ast::Intrinsic::StateClear
-        | sway_ast::Intrinsic::StateLoadWord
-        | sway_ast::Intrinsic::StateStoreWord
-        | sway_ast::Intrinsic::StateLoadQuad
-        | sway_ast::Intrinsic::StateStoreQuad
-        | sway_ast::Intrinsic::Log
-        | sway_ast::Intrinsic::Revert
-        | sway_ast::Intrinsic::Smo => Err(ConstEvalError::CannotBeEvaluatedToConst {
+            })),
+            (ConstantValue::U256(val1), ConstantValue::U256(val2)) => Ok(Some(Constant {
+                ty: Type::get_bool(lookup.context),
+                value: ConstantValue::Bool(val1 < val2),
+            })),
+            _ => {
+                unreachable!("Type checker allowed non integer value for LessThan")
+            }
+        },
+        Intrinsic::AddrOf
+        | Intrinsic::PtrAdd
+        | Intrinsic::PtrSub
+        | Intrinsic::IsReferenceType
+        | Intrinsic::IsStrType
+        | Intrinsic::Gtf
+        | Intrinsic::StateClear
+        | Intrinsic::StateLoadWord
+        | Intrinsic::StateStoreWord
+        | Intrinsic::StateLoadQuad
+        | Intrinsic::StateStoreQuad
+        | Intrinsic::Log
+        | Intrinsic::Revert
+        | Intrinsic::Smo => Err(ConstEvalError::CannotBeEvaluatedToConst {
             span: intrinsic.span.clone(),
         }),
-        sway_ast::Intrinsic::Not => {
-            // Not works only with uint at the moment
+        Intrinsic::Not => {
+            // Not works only with uint/u256 at the moment
             // `bool` ops::Not implementation uses `__eq`.
 
-            assert!(args.len() == 1 && args[0].ty.is_uint(lookup.context));
+            assert!(args.len() == 1);
```

Comment:
The type checking is actually done before const_eval is called, that is why we panic here in some cases. But I returned the asserts because they do make sense.